### PR TITLE
Enable the new cancellationToken pattern on non-azure libraries

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes https://github.com/Azure/autorest.csharp/issues/4243

# Description

Long story short: in non-azure libraries, convenience methods no longer have the parameter `cancellationToken`. Customers needing to provide a cancelling context should use the protocol methods instead for advanced scenarios.

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

**Comment by ArcturusZhang** [comment]
@m-nash @annelo-msft this PR is ready to review

File: test/CadlRanchProjectsNonAzure/authentication/http/custom/src/Generated/CustomClient.cs
Comment (annelo-msft) [inline test/CadlRanchProjectsNonAzure/authentication/http/custom/src/Generated/CustomClient.cs:99]: Non-blocking question: do we want the string "Azure" to appear here?  We could use an aka.ms link instead of the full path if we wanted to remove Azure from unbranded client code and docs.

**Review by annelo-msft — APPROVED** [review]
Approving with regard to the output code, which looks correct to me.  I defer to others on the generator implementation.

File: test/CadlRanchProjectsNonAzure/authentication/http/custom/src/Generated/CustomClient.cs
Comment (ArcturusZhang) [reply test/CadlRanchProjectsNonAzure/authentication/http/custom/src/Generated/CustomClient.cs:99]: I would change this to [aka.ms/azsdk/net/protocol-methods](https://aka.ms/azsdk/net/protocol-methods) as you suggested in https://github.com/annelo-msft/openai-in-typespec/pull/1#discussion_r1550115744

**Review by archerzz — APPROVED** [review]
LGTM.

File: test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs
Comment (annelo-msft) [inline test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs:1156]: Curious: why change the name of this method instead of making `RequestOptions` a required parameter to the protocol method?

Comment (ArcturusZhang) [reply test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs:1156]: I did not notice this change. I will investigate

Comment (ArcturusZhang) [reply test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs:1156]: There is another issue in this PR when similar things happen:
image
This method is calling itself, but it should call its corresponding protocol method

File: test/UnbrandedProjects/Customized-TypeSpec/src/Generated/SuperClient.cs
Comment (ArcturusZhang) [inline test/UnbrandedProjects/Customized-TypeSpec/src/Generated/SuperClient.cs:406]: this is terribly wrong, because this causes infinite invocation and leads to stack over flow.

**Comment by ArcturusZhang** [comment]
The doc changes are distractions, I will extract those changes into a standalone PR and merge it first.

File: test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs
Comment (ArcturusZhang) [reply test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs:1156]: as for this, this is a bug according to the document @pshao25 has wrote: https://github.com/Azure/azure-sdk-for-net/pull/37795/files
In this case, we should not have this convenience method at all, but something is wrong and we have this.
We will track this issue in another issue, since the problem is not introduced in this PR, this PR will not fix it.

**Comment by ArcturusZhang** [comment]
@annelo-msft FYI I updated this PR and now it only contains the change of removing cancellationToken (the other doc link change has been merged in another PR), it should be a lot more clear to review.

File: test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs
Comment (pshao25) [reply test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs:1156]: Though our principle is that we don't generate convenience method if there is no value (i.e. no models involved), this case makes the protocol method internal, so we have to still generated convenience method otherwise there would be no operations. Considering the potential ambiguity when one day user makes this protocol method public, we need to choose a way to handle it. Appending `Value` is the current behavior, we could adjust to make `RequestContext` required.

File: test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs
Comment (annelo-msft) [inline test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs:272]: Is this a recursive call now?  It may be that we still need to create an instance of RequestOptions in some cases, where the required RequestOptions is the only thing that differentiates a convenience method from a protocol method.

In this case, the correct thing to do here would either be:

```csharp
RequestOptions options = new();
ClientResult result = HelloDemo2(options);
```

or

```csharp
ClientResult result = HelloDemo2((RequestOptions)null);
```

File: test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs
Comment (annelo-msft) [reply test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs:1156]: I agree that in the case where the convenience method doesn't take an input model and doesn't return an output model it is unnecessary to generate the convenience method.  

For the image in https://github.com/Azure/autorest.csharp/pull/4524/files#r1556782904, the convenience method is necessary because, even though there is no input model, the convenience method needs to return an output model.  We should make RequestOptions required on the protocol method and the convenience method should pass either a new instance of RequestOptions to call the convenience method instead of itself, or null cast to RequestOptions.

I have illustrated the proposed fix in https://github.com/Azure/autorest.csharp/pull/4524/files#r1557834090.

File: test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs
Comment (ArcturusZhang) [reply test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs:272]: Yeah I am making a change to do the second.

**Comment by ArcturusZhang** [comment]
The merge broke this PR right now :( need to fix.

File: test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs
Comment (ArcturusZhang) [reply test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs:272]: I kind of get worried for issues like this, therefore I decided to change the generator so that when we call a protocol method here, we always pass all the parameters defined in the corresponding protocol method, no matter the parameter is optional or required (which should change nothing in run time, but this should make it more precise and acurate)

File: test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs
Comment (ArcturusZhang) [reply test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs:1164]: since the "append value suffix" case is not introduced by this PR, I will not change it here.
@pshao25 please feel free to open an issue to discuss that case when protocol is internal, and it does not have models as input or output, should we make the request context required, or append value to convenience method.

File: test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs
Comment (annelo-msft) [reply test/UnbrandedProjects/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs]: I like this approach -- I can't think of any issues with it offhand, and I agree it's nice to make it more explicit.  The only suggestion I'd make here, is that it can be confusing for readers of the code when we pass nulls in parameter lists -- if it's not too hard to add in the explicit argument name, that might make the code more readable and thereby more maintainable.  For example, instead of 
 
```
ClientResult result = AnonymousBody(content, null);
```
 
do:
 
```
ClientResult result = AnonymousBody(content, options: null);
```
 
But either way should work if that's too much work for the generator.

**Pull request merged**
